# 1.5 Запросы корректировки данных

---

## Инициализация БД

In [1]:
%load_ext sql
import sqlalchemy
engine = sqlalchemy.create_engine('mysql://user:pass@localhost:3306/stepik')
%sql mysql://user:pass@localhost:3306/stepik

'Connected: user@stepik'

## Заполнение таблицы из CSV файла

In [6]:
# Чтение файла в DataFrame
import pandas as pd
file = 'tables/book.csv'
df = pd.read_csv(file)
df

book_id                  title            author   price  amount
0        1     Мастер и Маргарита     Булгаков М.А.  670.99       3
1        2          Белая гвардия     Булгаков М.А.  540.50       5
2        3                  Идиот  Достоевский Ф.М.  460.00      10
3        4      Братья Карамазовы  Достоевский Ф.М.  799.01       2
4        5  Стихотворения и поэмы       Есенин С.А.  650.00      15

Создание схемы таблицы:

In [36]:
%%sql

  DROP TABLE IF EXISTS book;
CREATE TABLE IF NOT EXISTS book (
       book_id INT PRIMARY KEY AUTO_INCREMENT,
       title   VARCHAR(50),
       author  VARCHAR(30),
       price   DECIMAL(8, 2),
       amount  INT
);

 * mysql://user:***@localhost:3306/stepik
0 rows affected.
0 rows affected.


[]

In [37]:
# Запись данных в таблицу из DataFrame
types = {
    'book_id': sqlalchemy.Integer(),
    'price'  : sqlalchemy.Numeric(precision=8, scale=2),
    'amount' : sqlalchemy.Integer()
}
df.to_sql('book', con=engine, index=False, if_exists='append', dtype=types, method='multi')

***

# Упражнения

## Создание пустой таблицы `CREATE`

**Задание**  
Создать таблицу поставка (`supply`), которая имеет ту же структуру, что и таблиц `book`.

| Поле | Тип, описание |
|:---:|:---:|
| supply_id | INT PRIMARY KEY AUTO_INCREMENT |
| title | VARCHAR(50) |
| author | VARCHAR(30) |
| price | DECIMAL(8, 2) |
| amount | INT |

In [28]:
%%sql
CREATE TABLE supply (
       supply_id INT PRIMARY KEY AUTO_INCREMENT,
       title     VARCHAR(50),
       author    VARCHAR(30),
       price     DECIMAL(8,2),
       amount    INT
)

 * mysql://user:***@localhost:3306/stepik
(MySQLdb._exceptions.OperationalError) (1050, "Table 'supply' already exists")
[SQL: CREATE TABLE supply (
       supply_id INT PRIMARY KEY AUTO_INCREMENT,
       title     VARCHAR(50),
       author    VARCHAR(30),
       price     DECIMAL(8,2),
       amount    INT
)]
(Background on this error at: https://sqlalche.me/e/14/e3q8)


## Добавление записей в таблицу `INSERT`

**Задание**  
Занесите в таблицу `supply` четыре записи, чтобы получилась следующая таблица:

| supply_id | title | author | price | amount |
|:---:|:---:|:---:|:---:|:---:|
| 1 | Лирика | Пастернак Б.Л. | 518.99 | 2 |
| 2 | Черный человек  | Есенин С.А. | 570.20 | 6 |
| 3 | Белая гвардия | Булгаков М.А. | 540.50 | 7 |
| 4 | Идиот | Достоевский Ф.М. | 360.80 | 3 |

In [42]:
%%sql
INSERT INTO supply (title, author, price, amount)
VALUES ("Лирика", "Пастернак Б.Л.", 518.99, 2),
       ("Черный человек", "Есенин С.А.", 570.20, 6),
       ("Белая гвардия", "Булгаков М.А.", 540.50, 7),
       ("Идиот", "Достоевский Ф.М.", 360.80, 3)

 * mysql://user:***@localhost:3306/stepik
4 rows affected.


[]

In [43]:
%%sql
SELECT * FROM supply

 * mysql://user:***@localhost:3306/stepik
4 rows affected.


[(25, 'Лирика', 'Пастернак Б.Л.', Decimal('518.99'), 2),
 (26, 'Черный человек', 'Есенин С.А.', Decimal('570.20'), 6),
 (27, 'Белая гвардия', 'Булгаков М.А.', Decimal('540.50'), 7),
 (28, 'Идиот', 'Достоевский Ф.М.', Decimal('360.80'), 3)]

## Добавление записей из другой таблицы `INSERT` + `SELECT`

**Задание**  
Добавить из таблицы `supply` в таблицу `book`, все книги, кроме книг, написанных Булгаковым М.А. и Достоевским Ф.М.

In [9]:
%%sql
INSERT INTO book (title, author, price, amount)
SELECT title, author, price, amount
  FROM supply
 WHERE author NOT IN ('Булгаков М.А.', 'Достоевский Ф.М.');

SELECT * FROM book

 * mysql://user:***@localhost:3306/stepik
2 rows affected.
7 rows affected.


[(1, 'Мастер и Маргарита', 'Булгаков М.А.', Decimal('670.99'), 3),
 (2, 'Белая гвардия', 'Булгаков М.А.', Decimal('540.50'), 5),
 (3, 'Идиот', 'Достоевский Ф.М.', Decimal('460.00'), 10),
 (4, 'Братья Карамазовы', 'Достоевский Ф.М.', Decimal('799.01'), 2),
 (5, 'Стихотворения и поэмы', 'Есенин С.А.', Decimal('650.00'), 15),
 (6, 'Лирика', 'Пастернак Б.Л.', Decimal('518.99'), 2),
 (7, 'Черный человек', 'Есенин С.А.', Decimal('570.20'), 6)]

## Добавление записей, вложенные запросы

**Задание**  
Занести из таблицы `supply` в таблицу `book` только те книги, авторов которых нет в  `book`.

In [19]:
%%sql
INSERT INTO book (title, author, price, amount)
           SELECT title, author, price, amount
             FROM supply
            WHERE author NOT IN (
                      SELECT DISTINCT author
                        FROM book);

SELECT * FROM book

 * mysql://user:***@localhost:3306/stepik
1 rows affected.
6 rows affected.


[(1, 'Мастер и Маргарита', 'Булгаков М.А.', Decimal('670.99'), 3),
 (2, 'Белая гвардия', 'Булгаков М.А.', Decimal('540.50'), 5),
 (3, 'Идиот', 'Достоевский Ф.М.', Decimal('460.00'), 10),
 (4, 'Братья Карамазовы', 'Достоевский Ф.М.', Decimal('799.01'), 2),
 (5, 'Стихотворения и поэмы', 'Есенин С.А.', Decimal('650.00'), 15),
 (6, 'Лирика', 'Пастернак Б.Л.', Decimal('518.99'), 2)]

## Запросы на обновление `UPDATE`

Под обновлением данных подразумевается изменение значений в существующих записях таблицы.  
При этом возможно как изменение значений полей в группе строк (даже всех строк таблицы),  
так и правка значения поля отдельной строки.

Изменение записей в таблице реализуется с помощью запроса `UPDATE` .  
Простейший запрос на обновление выглядит так:  
> `UPDATE таблица SET поле = выражение`  

где:  
**таблица** –   имя таблицы, в которой будут проводиться изменения;  
**поле** –      поле таблицы, в которое будет внесено изменение;  
**выражение** – выражение,  значение которого будет занесено в поле.  

**Пример**  
Уменьшить на 30% цену книг в таблице `book`.

In [29]:
%%sql
UPDATE book 
   SET price = price * 0.7;
    
SELECT * FROM book

 * mysql://user:***@localhost:3306/stepik
5 rows affected.
5 rows affected.


[(1, 'Мастер и Маргарита', 'Булгаков М.А.', Decimal('469.69'), 3),
 (2, 'Белая гвардия', 'Булгаков М.А.', Decimal('378.35'), 5),
 (3, 'Идиот', 'Достоевский Ф.М.', Decimal('322.00'), 10),
 (4, 'Братья Карамазовы', 'Достоевский Ф.М.', Decimal('559.31'), 2),
 (5, 'Стихотворения и поэмы', 'Есенин С.А.', Decimal('455.00'), 15)]

---
С помощью запросов на обновление можно изменять не все записи в таблице (как в предыдущем запросе),  
а только часть из них.  
Для этого в запрос включается ключевое слово `WHERE` ,  
после которого указывается условие отбора строк для изменения.

**Пример**  
Уменьшить на 30% цену тех книг в таблице `book`, количество которых меньше 5.

In [32]:
%%sql
UPDATE book
   SET price = price * 0.7
 WHERE amount < 5;

SELECT * FROM book;

 * mysql://user:***@localhost:3306/stepik
2 rows affected.
5 rows affected.


[(1, 'Мастер и Маргарита', 'Булгаков М.А.', Decimal('469.69'), 3),
 (2, 'Белая гвардия', 'Булгаков М.А.', Decimal('540.50'), 5),
 (3, 'Идиот', 'Достоевский Ф.М.', Decimal('460.00'), 10),
 (4, 'Братья Карамазовы', 'Достоевский Ф.М.', Decimal('559.31'), 2),
 (5, 'Стихотворения и поэмы', 'Есенин С.А.', Decimal('650.00'), 15)]

**Задание**  
Уменьшить на 10% цену тех книг в таблице `book` ,  
количество которых принадлежит интервалу от 5 до 10, включая границы.

In [33]:
%%sql
UPDATE book
   SET price = price * 0.9
 WHERE amount BETWEEN 5 AND 10;

SELECT * FROM book

 * mysql://user:***@localhost:3306/stepik
2 rows affected.
5 rows affected.


[(1, 'Мастер и Маргарита', 'Булгаков М.А.', Decimal('469.69'), 3),
 (2, 'Белая гвардия', 'Булгаков М.А.', Decimal('486.45'), 5),
 (3, 'Идиот', 'Достоевский Ф.М.', Decimal('414.00'), 10),
 (4, 'Братья Карамазовы', 'Достоевский Ф.М.', Decimal('559.31'), 2),
 (5, 'Стихотворения и поэмы', 'Есенин С.А.', Decimal('650.00'), 15)]

## Запросы на обновление нескольких столбцов

Запросом `UPDATE` можно обновлять значения нескольких столбцов одновременно.  
В этом случае простейший запрос будет выглядеть так:  
> `UPDATE таблица SET поле1 = выражение1, поле2 = выражение2`

На складе, кроме хранения и получения книг, выполняется их оптовая продажа.  
Для реализации этого действия включим дополнительный столбец `buy`  в таблицу `book`:

| book_id | title | author | price | amount | buy |
|:---:|:---:|:---:|:---:|:---:|:---:|
| INT PRIMARY KEY AUTO_INCREMENT | VARCHAR(50) | VARCHAR(30) | DECIMAL(8,2) | INT | int |
| 1 | Мастер и Маргарита | Булгаков М.А. | 670.99 | 3 | 0 |
| 2 | Белая гвардия | Булгаков М.А. | 540.50 | 5 | 3 |
| 3 | Идиот | Достоевский Ф.М. | 460.00 | 10 | 8 |
| 4 | Братья Карамазовы | Достоевский Ф.М. | 799.01 | 2 | 0 |
| 5 | Стихотворения и поэмы | Есенин С.А. | 650.00 | 15 | 18 |

### Загрузка таблицы со столбцом `buy`

In [2]:
# Чтение файла в DataFrame
import pandas as pd
file = 'tables/book_with_buy.csv'
df = pd.read_csv(file)

In [3]:
%%sql

  DROP TABLE IF EXISTS book;
CREATE TABLE IF NOT EXISTS book (
       book_id INT PRIMARY KEY AUTO_INCREMENT,
       title   VARCHAR(50),
       author  VARCHAR(30),
       price   DECIMAL(8, 2),
       amount  INT,
       buy     INT
);

 * mysql://user:***@localhost:3306/stepik
0 rows affected.
0 rows affected.


[]

In [4]:
# Запись данных в таблицу из DataFrame
types = {
    'book_id': sqlalchemy.Integer(),
    'price'  : sqlalchemy.Numeric(precision=8, scale=2),
    'amount' : sqlalchemy.Integer(),
    'buy'    : sqlalchemy.Integer()
}
df.to_sql('book', con=engine, index=False, if_exists='append', dtype=types, method='multi')

**Пример**  
В столбце `buy` покупатель указывает количество книг, которые он хочет приобрести.  
Для каждой книги, выбранной покупателем, необходимо уменьшить ее количество на складе  
на указанное в столбцеbuy количество, а в столбец `buy` занести 0.

In [38]:
%%sql
UPDATE book
   SET amount = amount - buy,
       buy    = 0;
    
SELECT * FROM book

 * mysql://user:***@localhost:3306/stepik
5 rows affected.
5 rows affected.


[(1, 'Мастер и Маргарита', 'Булгаков М.А.', Decimal('670.99'), 3, 0),
 (2, 'Белая гвардия', 'Булгаков М.А.', Decimal('540.50'), 2, 0),
 (3, 'Идиот', 'Достоевский Ф.М.', Decimal('460.00'), 2, 0),
 (4, 'Братья Карамазовы', 'Достоевский Ф.М.', Decimal('799.01'), 2, 0),
 (5, 'Стихотворения и поэмы', 'Есенин С.А.', Decimal('650.00'), -3, 0)]

Как видно из таблицы, без проверки данных, которые занесены в столбец,  
нельзя запускать запрос на обновление (может получиться отрицательное значение количества).

**Задание**  
В таблице `book` необходимо скорректировать значение для покупателя в столбце `buy` таким образом,  
чтобы оно не превышало количество экземпляров книг, указанных в столбце `amount` .  
А цену тех книг, которые покупатель не заказывал, снизить на `10%`.

In [5]:
%%sql
UPDATE book
   SET buy   = IF(buy > amount, amount, buy),
       price = price * IF(buy = 0, 0.9, 1);
       
SELECT * FROM book

 * mysql://user:***@localhost:3306/stepik
5 rows affected.
5 rows affected.


[(1, 'Мастер и Маргарита', 'Булгаков М.А.', Decimal('603.89'), 3, 0),
 (2, 'Белая гвардия', 'Булгаков М.А.', Decimal('540.50'), 5, 3),
 (3, 'Идиот', 'Достоевский Ф.М.', Decimal('460.00'), 10, 8),
 (4, 'Братья Карамазовы', 'Достоевский Ф.М.', Decimal('719.11'), 2, 0),
 (5, 'Стихотворения и поэмы', 'Есенин С.А.', Decimal('650.00'), 15, 15)]

## Запросы на обновление нескольких таблиц

В запросах на обновление можно использовать несколько таблиц, но тогда

- для столбцов, имеющих одинаковые имена, необходимо указывать имя таблицы, к которой они относятся,  
  например, `book.price` – столбец price из таблицы `book`, `supply.price` – столбец `price` из таблицы `supply`
- все таблицы, используемые в запросе, нужно перечислить после ключевого слова `UPDATE`
- в запросе обязательно условие `WHERE`, в котором указывается условие при котором обновляются данные.

**Пример**  
Если в таблице `supply`  есть те же книги, что и в таблице `book`, добавлять эти книги в таблицу `book` не имеет смысла.  
Необходимо увеличить их количество на значение столбца `amount` таблицы `supply`.

In [11]:
%%sql
UPDATE book AS b, supply AS s
   SET b.amount = b.amount + s.amount
 WHERE b.title = s.title
   AND b.author = s.author;

SELECT * FROM book

 * mysql://user:***@localhost:3306/stepik
2 rows affected.
5 rows affected.


[(1, 'Мастер и Маргарита', 'Булгаков М.А.', Decimal('670.99'), 3),
 (2, 'Белая гвардия', 'Булгаков М.А.', Decimal('540.50'), 12),
 (3, 'Идиот', 'Достоевский Ф.М.', Decimal('460.00'), 13),
 (4, 'Братья Карамазовы', 'Достоевский Ф.М.', Decimal('799.01'), 2),
 (5, 'Стихотворения и поэмы', 'Есенин С.А.', Decimal('650.00'), 15)]

В этом запросе увеличилось количество двух книг: «Белая гвардия», которая в `supply` имеет ту же цену, и «Идиот», но цена этой книги в таблицах `book` и `supply` отличается. Для этой книги нужно пересчитать цену.

**Задание**  
Для тех книг в таблице `book` , которые есть в таблице `supply`,  
не только увеличить их количество в таблице `book` ( увеличить их количество на значение столбца `amount` таблицы `supply`),  
но и пересчитать их цену (для каждой книги найти сумму цен из таблиц `book` и `supply` и разделить на 2).

In [17]:
%%sql
UPDATE book AS b, supply AS s
   SET b.amount = b.amount + s.amount,
       b.price  = (b.price + s.price) / 2
 WHERE b.title  = s.title
   AND b.author = s.author;
    
SELECT * FROM book

 * mysql://user:***@localhost:3306/stepik
2 rows affected.
5 rows affected.


[(1, 'Мастер и Маргарита', 'Булгаков М.А.', Decimal('670.99'), 3),
 (2, 'Белая гвардия', 'Булгаков М.А.', Decimal('540.50'), 12),
 (3, 'Идиот', 'Достоевский Ф.М.', Decimal('410.40'), 13),
 (4, 'Братья Карамазовы', 'Достоевский Ф.М.', Decimal('799.01'), 2),
 (5, 'Стихотворения и поэмы', 'Есенин С.А.', Decimal('650.00'), 15)]

## Запросы на удаление `DELETE`

**Пример**  
После того, как информация о книгах из таблицы `supply` перенесена в `book` , необходимо очистить таблицу `supply`

In [41]:
%%sql
DELETE FROM supply;

SELECT * FROM supply

 * mysql://user:***@localhost:3306/stepik
3 rows affected.
0 rows affected.


[]

---
Запрос на удаления позволяет удалить не все записи таблицы, а только те,  
которые удовлетворяют условию, указанному после ключевого слова `WHERE`

**Пример**  
Удалить из таблицы `supply` все книги, названия которых есть в таблице `book`.

In [27]:
%%sql
DELETE FROM supply
 WHERE title IN (
         SELECT title
           FROM book
      );

SELECT * FROM supply

 * mysql://user:***@localhost:3306/stepik
2 rows affected.
2 rows affected.


[(9, 'Лирика', 'Пастернак Б.Л.', Decimal('518.99'), 2),
 (10, 'Черный человек', 'Есенин С.А.', Decimal('570.20'), 6)]

**Задание**  
Удалить из таблицы `supply` книги тех авторов, общее количество экземпляров книг которых в таблице `book` превышает 10.

In [44]:
%%sql
DELETE 
  FROM supply
 WHERE author IN (
          SELECT author
            FROM book
           GROUP BY author
          HAVING SUM(amount) > 10
       );

SELECT * FROM supply

 * mysql://user:***@localhost:3306/stepik
2 rows affected.
2 rows affected.


[(25, 'Лирика', 'Пастернак Б.Л.', Decimal('518.99'), 2),
 (27, 'Белая гвардия', 'Булгаков М.А.', Decimal('540.50'), 7)]

## Запросы на создание таблицы `CREATE` + `AS SELECT`

Новая таблица может быть создана на основе данных из другой таблицы.  
Для этого используется запрос `SELECT`, результирующая таблица которого и будет новой таблицей базы данных.  
При этом имена столбцов запроса становятся именами столбцов новой таблицы.  
Запрос на создание новой таблицы имеет вид:  

>`CREATE TABLE` _`имя_таблицы`_ `AS  
SELECT ...`

**Пример**  
Создать таблицу заказ (`ordering`), куда включить авторов и названия тех книг, количество экземпляров которых в таблице `book` меньше 4.  
Для всех книг указать одинаковое количество экземпляров - 5.

In [46]:
%%sql
CREATE TABLE ordering
    AS SELECT author, title, 5 AS amount
         FROM book
        WHERE amount < 4;
        
SELECT * FROM ordering

 * mysql://user:***@localhost:3306/stepik
2 rows affected.
2 rows affected.


[('Булгаков М.А.', 'Мастер и Маргарита', 5),
 ('Достоевский Ф.М.', 'Братья Карамазовы', 5)]

---
При создании таблицы можно использовать вложенные запросы как после `SELECT`, так и после `WHERE`.

**Пример**  
Создать таблицу заказ (`ordering`), куда включить авторов и названия тех книг,  
количество экземпляров которых в таблице `book` меньше 4.  
Для всех книг указать одинаковое значение - среднее количество экземпляров книг в таблице `book`.

In [48]:
%%sql
CREATE TABLE ordering
    AS SELECT author, title,
              (SELECT ROUND(AVG(amount))
                 FROM book) AS amount
         FROM book
        WHERE amount < 4;

SELECT * FROM ordering

 * mysql://user:***@localhost:3306/stepik
2 rows affected.
2 rows affected.


[('Булгаков М.А.', 'Мастер и Маргарита', Decimal('7')),
 ('Достоевский Ф.М.', 'Братья Карамазовы', Decimal('7'))]

**Задание**  
Создать таблицу заказ (`ordering`), куда включить авторов и названия тех книг, количество экземпляров которых в таблице `book` меньше среднего количества экземпляров книг в таблице `book` .  
В таблицу включить столбец `amount`, в котором для всех книг указать одинаковое значение - среднее количество экземпляров книг в таблице `book`.

In [50]:
%%sql
CREATE TABLE ordering
    AS SELECT author, title,
              (SELECT ROUND(AVG(amount))
                 FROM book) AS amount
         FROM book
        WHERE amount < (
              SELECT AVG(amount)
                FROM book);

SELECT * FROM ordering

 * mysql://user:***@localhost:3306/stepik
3 rows affected.
3 rows affected.


[('Булгаков М.А.', 'Мастер и Маргарита', Decimal('7')),
 ('Булгаков М.А.', 'Белая гвардия', Decimal('7')),
 ('Достоевский Ф.М.', 'Братья Карамазовы', Decimal('7'))]